fdbserver/DDTeamCollection.actor.cpp

```diff
@@ -195,13 +195,45 @@ class DDTeamCollectionImpl {
 		req.reply.send(std::make_pair(res, false));
 	}
 
+	// Return a threshold of storage queue size which guarantees at least DD_LONG_STORAGE_QUEUE_MAJORITY_PERCENTILE
+	// portion of storage servers that have longer storage queues
+	static Optional<int64_t> selectLongStorageQueueThreshold(const std::vector<Reference<TCTeamInfo>>& teams) {
+		std::vector<int64_t> queueLengthList;
+		for (int i = 0; i < teams.size(); i++) {
+			Optional<int64_t> storageQueueSize = teams[i]->getLongestStorageQueueSize();
+			if (!storageQueueSize.present()) {
+				// This SS may not healthy, so avoid selecting them by setting the queue length to a large number
+				queueLengthList.push_back(std::numeric_limits<int64_t>::max());
+			} else {
+				queueLengthList.push_back(storageQueueSize.get());
+			}
+		}
+		if (queueLengthList.size() <
+		    SERVER_KNOBS->DD_MIN_TEAM_PERCENTAGE_TO_SELECT_LONG_QUEUE_THRESHOLD * teams.size()) {
+			TraceEvent(SevWarn, "StorageQueueAwareGetThresholdQueueFailed")
+			    .suppressFor(5.0)
+			    .detail("AvailableTeamCount", queueLengthList.size());
+			return Optional<int64_t>();
+		}
+		std::sort(queueLengthList.begin(), queueLengthList.end());
```

Comment:
consider using `std::nth_element` to only partially sort the list, may be more efficient.
```
std::nth_element(queueLengthList.begin(), queueLengthList.begin() + position, queueLengthList.end());

int64_t threshold = queueLengthList[position];
```